gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java

```diff
@@ -695,12 +696,25 @@ public void write(JsonWriter out, Locale value) throws IOException {
   public static final TypeAdapterFactory LOCALE_FACTORY = newFactory(Locale.class, LOCALE);
 
   public static final TypeAdapter<JsonElement> JSON_ELEMENT = new TypeAdapter<JsonElement>() {
-    @Override public JsonElement read(JsonReader in) throws IOException {
-      if (in instanceof JsonTreeReader) {
-        return ((JsonTreeReader) in).nextJsonElement();
+    /**
+     * Tries to begin reading a JSON array or JSON object, returning {@code null} if
+     * the next element is neither of those.
+     */
+    private JsonElement tryBeginNesting(JsonReader in, JsonToken peeked) throws IOException {
+      if (peeked == JsonToken.BEGIN_ARRAY) {
```

Comment:
Same comment about switch here, and about `LinkedList` below. Is it possible to reduce the amount of code duplication between this class and `ObjectTypeAdapter`?

Reply:
> Is it possible to reduce the amount of code duplication between this class and `ObjectTypeAdapter`?

The code indeed looks pretty similar. Maybe this can be done by having a separate internal abstract `IterativeDeserializer` or similar with subclasses for `Object` and `JsonElement` deserialization, but I am not sure if this wouldn't be more complex (and possibly more inefficient). Do you have a specific implementation in mind?

Reply:
I don't think it's necessary to do anything about this right away, unless there is some way that people could switch between the two to circumvent any depth (etc) limitations we are trying to introduce here.